crates/chain/src/local_chain.rs

```diff
@@ -17,224 +152,339 @@ impl ChainOracle for LocalChain {
+            changeset.extend(
+                updated_cps
+                    .iter()
+                    .map(|(height, cp)| (*height, Some(cp.hash()))),
+            );
+            changeset
+        };
+
+        let apply_update = move || {
+            if let Some(&start_height) = updated_cps.keys().next() {
+                self.checkpoints.split_off(&invalidate_lb);
+                self.checkpoints.append(&mut updated_cps);
+                if !self.is_empty() && !complete_match {
+                    self.fix_links(start_height);
                 }
-                self.blocks
-                    .range(first_invalid_height..)
-                    .map(|(height, _)| (*height, None))
-                    .collect()
             }
-            None => BTreeMap::new(),
         };
-        for (height, update_hash) in update {
-            let original_hash = self.blocks.get(height);
-            if Some(update_hash) != original_hash {
-                changeset.insert(*height, Some(*update_hash));
-            }
-        }
 
-        Ok(changeset)
+        Ok((apply_update, changeset))
     }
 
-    /// Applies the given `changeset`.
-    pub fn apply_changeset(&mut self, changeset: ChangeSet) {
-        for (height, blockhash) in changeset {
-            match blockhash {
-                Some(blockhash) => self.blocks.insert(height, blockhash),
-                None => self.blocks.remove(&height),
-            };
+    /// Apply the given `changeset`.
+    pub fn apply_changeset(&mut self, changeset: &ChangeSet) {
+        if let Some(start_height) = changeset.keys().next().cloned() {
+            for (&height, &hash) in changeset {
+                match hash {
+                    Some(hash) => self
+                        .checkpoints
+                        .insert(height, CheckPoint::new(BlockId { height, hash })),
+                    None => self.checkpoints.remove(&height),
+                };
+            }
+            self.fix_links(start_height);
         }
     }
 
-    /// Updates [`LocalChain`] with an update [`LocalChain`].
+    /// Update [`LocalChain`].
     ///
     /// This is equivalent to calling [`determine_changeset`] and [`apply_changeset`] in sequence.
     ///
     /// [`determine_changeset`]: Self::determine_changeset
     /// [`apply_changeset`]: Self::apply_changeset
-    pub fn apply_update(&mut self, update: Self) -> Result<ChangeSet, UpdateNotConnectedError> {
-        let changeset = self.determine_changeset(&update)?;
-        self.apply_changeset(changeset.clone());
+    pub fn apply_update(&mut self, new_tip: CheckPoint) -> Result<ChangeSet, CannotConnectError> {
+        let (apply, changeset) = self.update(new_tip)?;
+        apply();
         Ok(changeset)
     }
 
+    /// Get or insert a `block_id`.
+    ///
+    /// # Errors
+    ///
+    /// Replacing the block hash of an existing checkpoint will result in an error.
+    pub fn get_or_insert(
```

Comment:
What's wrong with `get_block_hash` and `insert_block`. Why does this API need to exist?

Reply:
Reverted back to `insert_block`!